crypto/Pos/Crypto/Orphans.hs

```diff
@@ -8,51 +8,149 @@ import           Universum
+instance Hashable Ed25519.PublicKey where
+  hashWithSalt salt pk = hashWithSalt salt $ (BA.convert pk :: BS.ByteString)
+
+instance Hashable Ed25519.SecretKey where
+  hashWithSalt salt pk = hashWithSalt salt $ (BA.convert pk :: BS.ByteString)
+
+instance Hashable Ed25519.Signature where
+  hashWithSalt salt pk = hashWithSalt salt $ (BA.convert pk :: BS.ByteString)
+
+
+instance Ord Ed25519.PublicKey where
+  compare x1 x2 = compare (BA.convert x1 :: BS.ByteString) (BA.convert x2 :: BS.ByteString)
+
+instance Ord Ed25519.SecretKey where
+  compare x1 x2 = compare (BA.convert x1 :: BS.ByteString) (BA.convert x2 :: BS.ByteString)
+
+instance Ord Ed25519.Signature where
+  compare x1 x2 = compare (BA.convert x1 :: BS.ByteString) (BA.convert x2 :: BS.ByteString)
+
+
+instance SafeCopy BA.Bytes where
+  putCopy s =
+    let
+      toByteString :: BA.Bytes -> BS.ByteString
+      toByteString = BA.convert
+    in
+      contain $ safePut (toByteString s)
+  getCopy =
+    let
+      fromByteString :: BS.ByteString -> BA.Bytes
+      fromByteString = BA.convert
+    in
+      contain $ fromByteString <$> safeGet
+
+
+instance SafeCopy BA.ScrubbedBytes where
+  putCopy s =
+    let
+      toByteString :: BA.ScrubbedBytes -> BS.ByteString
+      toByteString = BA.convert
+    in
+      contain $ safePut (toByteString s)
+  getCopy =
+    let
+      fromByteString :: BS.ByteString -> BA.ScrubbedBytes
+      fromByteString = BA.convert
+    in
+      contain $ fromByteString <$> safeGet
 
-instance NFData Ed25519.PublicKey
-instance NFData Ed25519.SecretKey
-instance NFData Ed25519.Signature
 
 deriveSafeCopySimple 0 'base ''Ed25519.PublicKey
 deriveSafeCopySimple 0 'base ''Ed25519.SecretKey
 deriveSafeCopySimple 0 'base ''Ed25519.Signature
 
+fromByteString :: BS.ByteString -> BA.Bytes
+fromByteString = BA.convert
+
 instance FromJSON Ed25519.PublicKey where
-    parseJSON v = Ed25519.PublicKey . getJsonByteString <$> parseJSON v
+  parseJSON v =
+    let
+      fromCryptoToAeson :: CryptoFailable a -> Parser a
+      fromCryptoToAeson (CryptoFailed e) = error $ mappend "Pos.Crypto.Orphan.parseJSON Ed25519.PublicKey failed because " (T.pack $ show e)
+      fromCryptoToAeson (CryptoPassed r) = return r
+    in
+      do
+        res <- Ed25519.publicKey . fromByteString . getJsonByteString <$> parseJSON v
+        fromCryptoToAeson res
 
 instance ToJSON Ed25519.PublicKey where
-    toJSON = toJSON . JsonByteString . Ed25519.openPublicKey
+  toJSON =
+    let
+      toByteString :: Ed25519.PublicKey -> BS.ByteString
+      toByteString = BA.convert
+    in
+      toJSON . JsonByteString . toByteString
 
 instance FromJSON Ed25519.Signature where
-    parseJSON v = Ed25519.Signature . getJsonByteString <$> parseJSON v
+  parseJSON v =
+    let
+      fromCryptoToAeson :: CryptoFailable a -> Parser a
+      fromCryptoToAeson (CryptoFailed e) = error $ mappend "Pos.Crypto.Orphan.parseJSON Ed25519.Signature failed because " (T.pack $ show e)
```

Comment:
This function has been repeated twice now, so it should be factored into a top level helper.